fs-storage/src/file_storage.rs

```diff
@@ -0,0 +1,246 @@
+                for line in lines {
+                    let line = line?;
+                    if line.is_empty() {
+                        continue;
+                    }
+
+                    let parts: Vec<&str> =
+                        line.split(KEY_VALUE_SEPARATOR).collect();
+                    let id = K::from_str(parts[0])?;
+                    let value = V::from_str(parts[1])?;
+                    value_by_id.insert(id, value);
+                }
+
+                Ok(value_by_id)
+            }
+            Some(Err(e)) => Err(e.into()),
+            None => Err(ArklibError::Storage(
+                self.label.clone(),
+                "Storage file is missing header".to_owned(),
+            )),
+        }
+    }
+
+    /// Write data to file
+    ///
+    /// Data is a key-value mapping between [ResourceId] and a generic Value
+    pub fn write_file<K, V>(
+        &mut self,
+        value_by_id: &HashMap<K, V>,
+    ) -> Result<()>
+    where
+        K: Display,
+        V: Display,
+    {
+        fs::create_dir_all(self.path.parent().unwrap())?;
+        let file = File::create(&self.path)?;
+        let mut writer = BufWriter::new(file);
+
+        writer.write_all(
+            format!("{}{}\n", STORAGE_VERSION_PREFIX, STORAGE_VERSION)
+                .as_bytes(),
+        )?;
+
+        for (id, value) in value_by_id {
+            writer.write_all(
+                format!("{}{}{}\n", id, KEY_VALUE_SEPARATOR, value).as_bytes(),
+            )?;
+        }
+
+        let new_timestamp = fs::metadata(&self.path)?.modified()?;
+        if new_timestamp == self.timestamp {
+            return Err("Timestamp didn't update".into());
+        }
+        self.timestamp = new_timestamp;
+
+        log::info!(
+            "{} {} entries has been written",
+            self.log_prefix,
+            value_by_id.len()
+        );
+        Ok(())
+    }
+
+    /// Verify the version stored in the file header
+    fn verify_version(&self, header: &str) -> Result<()> {
+        if !header.starts_with(STORAGE_VERSION_PREFIX) {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Unknown storage version".to_owned(),
+            ));
+        }
+
+        let version = header[STORAGE_VERSION_PREFIX.len()..]
+            .parse::<i32>()
+            .unwrap();
+
+        match version.cmp(&STORAGE_VERSION) {
+            std::cmp::Ordering::Equal => Ok(()),
+            std::cmp::Ordering::Less => Err(ArklibError::Storage(
+                self.label.clone(),
+                "Storage format is newer than the app".to_owned(),
+            )),
+            std::cmp::Ordering::Greater => Err(ArklibError::Storage(
+                self.label.clone(),
+                "Storage format is older than the app".to_owned(),
+            )),
+        }
+    }
+}
+
+impl Drop for FileStorage {
+    fn drop(&mut self) {
+        if let Err(e) = fs::remove_file(&self.path) {
```

Comment:
Does this erases the file from disk every time we discard the storage object? The idea is to persist the storage on disk between runs.

Reply:
Aah this comes from the erase function implemented in Kotlin. But I guess instead of deleting the file by itself, it can be exposed as a function that can be called by the user.

Reply:
Yep, we need erasing function in the API but must not call it automatically.